### PR TITLE
[CMake] Add dataframe option

### PR DESCRIPTION
**Pull request by stwunsch**
Because RDataFrame is not available on 32bit, we have to disable the
experimental parts of TMVA which are dependent on it as well.

### CODE REVIEW

**Comment by amadio** [comment]
It may be a good idea to make dataframe optional and disable (or require it to be off) on 32bits machines. The proliferation of these things not availble on 32bits will only get more and more complicated. We already have a few other places where this needs to be checked. It's better to have a `dataframe` option than keeping in sync all the 32bit checks everywhere (i.e. which tests to run, which tutorials to veto, which optional components that depend on dataframe and cannot be build as a consequence, etc).

**Comment by stwunsch** [comment]
@Axel-Naumann correct me if I'm wrong. We do not plan to make dataframe optional since we expect gcc/clang to figure out their discrepancies and resolve the 32bit issue for RDF?

However, from the technical side, I agree with @amadio having a dataframe flag would make stuff much easier and for sure more consistent on the long run.

**Comment by bluehood** [comment]
@dpiparo I thought rdf was reenabled on 32 bits at some point, do I remember incorrectly?

@amadio I don't have anything against a dataframe option that is usually on and is turned off on 32 bit machines.

If I had time I'd check that we actually still need to disable rdf on 32 bits, but the failure was not easy to reproduce.

**Comment by stwunsch** [comment]
@amadio I put a commit on top with a dataframe option. Better like this?

**Comment by stwunsch** [comment]
Oh, well, I forgot some places.

File: cmake/modules/RootBuildOptions.cmake [hunk 1]
```diff
@@ -104,6 +104,7 @@ ROOT_BUILD_OPTION(cocoa OFF "Use native Cocoa/Quartz graphics backend (MacOS X o
 ROOT_BUILD_OPTION(coverage OFF "Enable compile flags for coverage testing")
 ROOT_BUILD_OPTION(cuda OFF "Enable support for CUDA (requires CUDA toolkit >= 7.5)")
 ROOT_BUILD_OPTION(cxxmodules OFF "Enable support for C++ modules")
+ROOT_BUILD_OPTION(dataframe ON "Enable ROOT dataframes")
```
Comment (amadio) [inline cmake/modules/RootBuildOptions.cmake]:
```suggestion
ROOT_BUILD_OPTION(dataframe ON "Enable ROOT RDataFrame")
```

File: tmva/tmva/CMakeLists.txt [hunk 2]
```diff
@@ -368,8 +364,26 @@ ROOT_STANDARD_LIBRARY_PACKAGE(TMVA
...
+    )
+endif()
+
+ROOT_STANDARD_LIBRARY_PACKAGE(TMVA
+  HEADERS ${TMVA_HEADERS}
+  SOURCES ${TMVA_SOURCES}
+  DICTIONARY_OPTIONS -writeEmptyRootPCM ${dictoptions}
+  DEPENDENCIES ${TMVA_DEPENDENCIES}
+  INSTALL_OPTIONS ${installoptions}
 )
```
Comment (amadio) [inline tmva/tmva/CMakeLists.txt]: Please set variables only for optional extras (look in `tree/dataframe/CMakeLists.txt` for an example). The above goes against the current coding style.

Comment (stwunsch) [reply tmva/tmva/CMakeLists.txt]: Thanks, done!

File: cmake/modules/RootBuildOptions.cmake (hunk 1 above)
Comment (stwunsch) [reply cmake/modules/RootBuildOptions.cmake]: Done

**Comment by stwunsch** [comment]
@phsft-bot build

**Comment by stwunsch** [comment]
@amadio @Axel-Naumann ok, that is getting out of hands :sweat_smile: we can have the very first commit to just fix the issue OR add the dataframe option to the build system, which is really a lot of changes.

Is just fixing the 32bit issue it worth?

**Comment by Axel-Naumann** [comment]
We should discuss that in person, given that this is easily possible :-) Can we do that on Monday?

**Comment by amadio** [comment]
As I mentioned on mattermost, the proliferation of bug prone if conditions is also getting out of hand and will be a nightmare to maintain. I think that we should reconsider this way of disabling RDataFrame and just re-enable it (and other deps) everywhere.

**Comment by stwunsch** [comment]
This is connected to the roottest PR [here](https://github.com/root-project/roottest/pull/339).

**Comment by stwunsch** [comment]
But hey, it works now :)

File: tmva/tmva/CMakeLists.txt [hunk 3]
```diff
@@ -368,8 +374,9 @@ ROOT_STANDARD_LIBRARY_PACKAGE(TMVA
     Imt
     RIO
     XMLIO
-  INSTALL_OPTIONS
-    ${installoptions}
+    ${TMVA_EXTRA_DEPENDENCIES}
+  DICTIONARY_OPTIONS -writeEmptyRootPCM ${dictoptions}
+  INSTALL_OPTIONS ${installoptions}
```
Comment (amadio) [inline tmva/tmva/CMakeLists.txt]: Please don't change the coding style. One thing per line. It's for better diffs when we make changes.

Comment (stwunsch) [reply tmva/tmva/CMakeLists.txt]: I really need to look up the coding style rules :see_no_evil:

**Comment by amadio** [comment]
Much nicer! I think with a couple of tweaks this will be ready to merge.

File: tmva/tmva/CMakeLists.txt (hunk 3 above)
Comment (stwunsch) [reply tmva/tmva/CMakeLists.txt]: Fixed.

**Comment by stwunsch** [comment]
@phsft-bot build

File: tmva/tmva/CMakeLists.txt (hunk 3 above)
Comment (amadio) [reply tmva/tmva/CMakeLists.txt]: I don't think we have it anywhere yet for CMake, but most files already follow the new style, so just keeping it consistent is enough.

File: config/root-config.in [hunk 4]
```diff
@@ -76,11 +76,10 @@ rootlibs="-lCore -lImt -lRIO -lNet -lHist -lGraf -lGraf3d -lGpad -lROOTVecOps -l
           -lRint -lPostscript -lMatrix -lPhysics -lMathCore -lThread\
           -lMultiProc"
 
-if [ "@UNIX@" != "1" ] || [ "@CMAKE_SIZEOF_VOID_P@" != "4" ]; then
+if [ "@dataframe@" == "ON" ]; then
```
Comment (amadio) [inline config/root-config.in]: We can use the suggestion below for now, but we should have a generic solution if more things are allowed to be OFF in the future:
```suggestion
if echo "${features}" | grep -q "dataframe"; then
```

Comment (Axel-Naumann) [reply config/root-config.in]: Meh, Stefan didn't enable pushing by repo owners. I'll merge (after amending) by hand.

**Comment by Axel-Naumann** [comment]
Merged as 31292b90826fbb1aaf0694c445f5996cdde3f358 by hand, with @amadio 's suggested change amended.

**Pull request closed without merging**